plutus-contract/src/Wallet/Emulator/Wallet.hs

```diff
@@ -459,22 +459,30 @@ handleBalanceTx utxo utx = do
 
     collateral <- traverse lookupValue (Tx.txCollateralInputs txWithinputsAdded)
 
-    let collAddr = maybe ownAddr Ledger.txOutAddress $ Tx.txReturnCollateral txWithinputsAdded
-        collateralPercent = maybe 100 fromIntegral (protocolParamCollateralPercent pProtocolParams)
-        collFees = Ada.toValue $ (Ada.fromValue fees * collateralPercent + 99 {- make sure to round up -}) `Ada.divide` 100
-        collBalance = fold collateral PlutusTx.- collFees
-
-    ((negColl, newTxInsColl), (_, mNewTxOutColl)) <- calculateTxChanges collAddr outRefsWithValue $ Value.split collBalance
-
-    txWithCollateralInputs <- if Value.isZero negColl
-        then do
-            logDebug NoCollateralInputsAdded
-            pure txWithinputsAdded
-        else do
-            logDebug $ AddingCollateralInputsFor negColl
-            pure $ txWithinputsAdded & over Tx.collateralInputs (sort . (++) (fmap Tx.pubKeyTxInput newTxInsColl))
-
-    pure $ txWithCollateralInputs & Tx.totalCollateral ?~ collFees & Tx.returnCollateral .~ mNewTxOutColl
+    if Value.isZero (fold collateral)
+        && null (Tx.txScripts txWithinputsAdded)
+        && null (Tx.txReferenceInputs txWithinputsAdded)
```

Comment:
I guess you _can_ reference a txout that _doesn't_ have an reference script. So this condition needs to be a bit different?